Source/OCMock/NSInvocation+OCMAdditions.m

```diff
@@ -41,16 +43,74 @@ + (NSInvocation *)invocationForBlock:(id)block withArguments:(NSArray *)argument
 }
 
 
-- (BOOL)hasCharPointerArgument
+static NSString *const OCMRetainedObjectArgumentsKey = @"OCMRetainedObjectArgumentsKey";
+
+- (void)retainObjectArgumentsExcluding:(id)objectToExclude
 {
-    NSMethodSignature *signature = [self methodSignature];
-    for(NSUInteger i = 0; i < [signature numberOfArguments]; i++)
+    if(objc_getAssociatedObject(self, OCMRetainedObjectArgumentsKey) == nil)
     {
-        const char *argType = OCMTypeWithoutQualifiers([signature getArgumentTypeAtIndex:i]);
-        if(strcmp(argType, "*") == 0)
-            return YES;
+        NSMutableArray *retainedArguments = [[NSMutableArray alloc] init];
+        NSMethodSignature *methodSignature = self.methodSignature;
+        
+        id target = self.target;
+        if((target != nil) && (target != objectToExclude) && !object_isClass(target))
+        {
+            // Bad things will happen if the target is a block since it's not being
+            // copied. There isn't a very good way to tell if an invocation's target
+            // is a block though (the argument type at index 0 is always "@" even if
+            // the target is a Class or block), and in practice it's OK since you
+            // can't mock a block.
+            [retainedArguments addObject:target];
```

Comment:
I gave an implementation to check whether an object is a block in the comments to this pull request... it's probably safer to test the object directly rather than rely on the types, I'd guess.